test/run_test.py

```diff
@@ -340,6 +349,82 @@ def print_to_stderr(message):
+    commit_date_str = subprocess.check_output(
+        ['git', 'show', '-s', '--format=%ci', 'HEAD'],
+        encoding="ascii").splitlines()[0]
+    commit_date = datetime.strptime(commit_date_str, '%Y-%m-%d %H:%M:%S %z')
+    day_before_commit = str(commit_date - timedelta(days=1)).split(' ')[0]
+    # something like git rev-list --before="2021-02-24" --max-count=1 --remotes="*upstream/nightly"
+    nightly_commit = subprocess.check_output(
+        ["git", "rev-list", f"--before={day_before_commit}", "--max-count=1", "--remotes=*upstream/nightly"],
+        encoding="ascii").splitlines()[0]
+
+    job = os.environ.get("CIRCLE_JOB", "")
+    job_minus_shard_number = job.rstrip('0123456789')
+    s3 = boto3.resource("s3", config=botocore.config.Config(signature_version=botocore.UNSIGNED))
+    bucket = s3.Bucket(name="ossci-metrics")
+    summaries = bucket.objects.filter(Prefix=f"test_time/{nightly_commit}/{job_minus_shard_number}")
+
+    reports = []
+    for summary in summaries:
+        binary = summary.get()["Body"].read()
+        string = bz2.decompress(binary).decode("utf-8")
+        reports.append(json.loads(string))
+    return reports
+
+
+def calculate_job_times(reports) -> Dict[str, Tuple[float, int]]:
+    # an entry will be like ("test_file_name" -> (current_avg, # values))
+    jobs_to_times: Dict[str, Tuple[float, int]] = dict()
+    for report in reports:
+        assert report['format_version'] == 2, "S3 format currently handled is version 2 only"
+        files = report['files']
+        for name, test_file in files.items():
+            if name != jobs_to_times:
+                jobs_to_times[name] = (test_file['total_seconds'], 1)
+            else:
+                curr_avg, curr_count = jobs_to_times[name]
+                new_count = curr_count + 1
+                new_avg = (curr_avg * curr_count + test_file['total_seconds']) / new_count
+                jobs_to_times[name] = (new_avg, new_count)
+    return jobs_to_times
+
+
+def calculate_shards(num_shards: int, tests: List[str], job_times: Dict[str, Tuple[float, int]]) -> List[Tuple[float, List[str]]]:
+    # if there's 'test_cpp_extensions_aot' entry in job_times, add 'test_cpp_extensions_aot_ninja' 
+    # and 'test_cpp_extensions_aot_no_ninja' duplicate entries to ease future computation since 
+    # test_cpp_extensions_aot_no_ninja and test_cpp_extensions_aot_ninja are Python test jobs that 
+    # both use the test_cpp_extensions_aot.py file.
+    if 'test_cpp_extensions_aot' in job_times:
+        job_times['test_cpp_extensions_aot_ninja'] = job_times['test_cpp_extensions_aot']
+        job_times['test_cpp_extensions_aot_no_ninja'] = job_times['test_cpp_extensions_aot']
+    filtered_job_times = {job: avg_time for job, (avg_time, _) in job_times.items() if job in tests}
+    sorted_jobs = sorted(filtered_job_times, key=lambda j: job_times[j], reverse=True)
+    unknown_jobs = [job for job in tests if job not in sorted_jobs]
```

Comment:
same here (time complexity is proportional to `len(tests) * len(sorted_jobs)` if we don't convert `sorted_jobs` to a set)